### DISCUSSION TITLE
Optional form field not working with test client

### DISCUSSION THREAD

**Discussion by ghost** [fastapi/fastapi#12227] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Annotated, Literal, Optional

from fastapi import FastAPI, Form
from fastapi.testclient import TestClient

app = FastAPI()

@app.post("/")
async def read_main(
    attribute : Annotated[Optional[Literal["abc", "def"]], Form()]
):
    print(attribute)


client = TestClient(app)


data = {}
data["attribute"] = None

response = client.post("/", data=data)
print(response.content)
assert response.status_code == 200
```

### Description

In FastAPI 0.113 .0 the above code passes.
Since version 0.114.0 the response is a 422 status code with error message '{"detail":[{"type":"literal_error","loc":["body","attribute"],"msg":"Input should be \'abc\' or \'def\'","input":"","ctx":{"expected":"\'abc\' or \'def\'"}}]}'

### Operating System

macOS

### FastAPI Version

0.114.0

### Pydantic Version

2.9.1

### Python Version

3.12.3

**Comment by TomFaulkner** [comment]
I think this is the same issue as here: https://github.com/fastapi/fastapi/discussions/12150#discussioncomment-10577318

**Reply by ghost** [reply]
The main differences I see:
This issue is with version 0.114, the linked one with 0.113, where my text example works just fine.
Also I do not use an explicit pydantic model.

**Comment by Kludex** [comment, marked as answer]
With FastAPI 0.113.0:

```py
{'detail': [{'type': 'missing', 'loc': ['body', 'attribute'], 'msg': 'Field required', 'input': None}]}
Traceback (most recent call last):
  File "/Users/marcelotryle/dev/encode/starlette/main.py", line 20, in <module>
    assert response.status_code == 200
AssertionError
```

With FastAPI 0.114.0:
```py
{
│   'detail': [
│   │   {
│   │   │   'type': 'literal_error',
│   │   │   'loc': ['body', 'attribute'],
│   │   │   'msg': "Input should be 'abc' or 'def'",
│   │   │   'input': '',
│   │   │   'ctx': {'expected': "'abc' or 'def'"}
│   │   }
│   ]
}
Traceback (most recent call last):
  File "/Users/marcelotryle/dev/encode/starlette/main.py", line 20, in <module>
    assert response.status_code == 200
AssertionError
```

There's a difference on behavior, but the "test" doesn't pass anyway... 

When you add a default value to `attribute`, it does pass:

```py
from typing import Annotated, Literal, Optional

from fastapi import FastAPI, Form
from fastapi.testclient import TestClient
from rich.pretty import pprint

app = FastAPI()


@app.post("/")
async def read_main(attribute: Annotated[Optional[Literal["abc", "def"]], Form()] = None):
    print(attribute)


client = TestClient(app)


response = client.post("/")
pprint(response.json())  # pragma: no branch
assert response.status_code == 200
```

In any case, I'll create an issue from this because I can see a change in behavior...

**Reply by Kludex** [reply]
- Follow https://github.com/fastapi/fastapi/issues/12245.

**Reply by ghost** [reply]
Hi, you are right, I forgot to add a default to my example, hence it didn't pass.
The actual issue I am seeing in my app is that I need to get multiple fields from the form, some of which can be optional.
Hence why I added the `data` to my sample request.

E.g. expanding your sample:
```
from typing import Annotated, Literal, Optional

from fastapi import FastAPI, Form
from fastapi.testclient import TestClient
from rich.pretty import pprint

app = FastAPI()


@app.post("/")
async def read_main(attribute_1: Annotated[int, Form()],
                    attribute_2: Annotated[Optional[Literal["abc", "def"]], Form()] = None):
    print(attribute_1)
    print(attribute_2)


client = TestClient(app)

data = {}
data["attribute_1"] = 123
data["attribute_2"] = None

response = client.post("/", data=data)
pprint(response.json())  # pragma: no branch
assert response.status_code == 200
```

will actually fail in 0.114 but pass in 0.113

One can remove setting "attribute_2" to None to make it pass again.

**Comment by christiancons** [comment]
Hi everyone,

we ran into a samilar issue and were able narrow the issue down to the behaviour of the TestClient.

see: https://github.com/fastapi/fastapi/issues/12245#issuecomment-2399589746
